Adding a code for editing in the Solr.

Romx98 opened the pull request.

Romx98 committed:
Summary description of changes:
 - added classes for update VC in Solr

Romx98 committed:
Summary description of changes:
 - added implementation Solr to gradle

Romx98 committed:
Summary description of changes:
 - added printSolrResponse,
 -

Romx98 committed:
Summary description of changes:
 - added new class SolrClientApi
   - contains general properties of solr (connection, query, commit and close)
   - designed for future use
 - modify SolrVcRecords, added method for removing the VC field
   - from this class have been moved to class SolrClientApi
 - !! ONLY removal VC !!
 - tested on the local system, everything is running OK...

kazooo commented on build.gradle:
```diff
@@ -14,6 +14,8 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.apache.solr:solr-solrj:8.7.0'
```
snaz se pouzivat posledni verze knihoven, pokud pises neco od zacatku
tady oprav na posledni verzi SolrJ

kazooo commented on src/main/java/cz/mzk/fedora/model/DataStreams.java:
```diff
@@ -4,7 +4,7 @@
 public enum DataStreams {
 
-    RELS_EXT("RELS-EXT", "application/rdf+xml" , "true", "X", "A");
+    RELS_EXT("RELS-EXT", "text/xml" , "true", "X", "A");
```
proc "text/xml"?

kazooo replied:
proc to tady potrebujes menit kdyz resis Solr?

Romx98 replied:
Toto som menil este pred spustenim skriptu na fedoru, lebo pozeral som si predosle verzie a tie to mali nastavene... a zabudol som to pred tym commitnut

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/RemoveVcSolr.java:
```diff
@@ -0,0 +1,22 @@
+public class RemoveVcSolr {
+
+    public static void main(String[] args) throws SolrServerException, IOException {
+        String solrHost = System.getenv("SH");
```
kdyz pristi nekde nekam budes davat prilis strucne env promenne, tak to muze zpusobit neporozumeni ze strany kolegu
treba ja bych nepochopil nekde promennou "SH", ale s napriklad "SOLR_HOST" je vsechno jasne)

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientApi.java:
```diff
@@ -0,0 +1,64 @@
+            SolrDocumentList docs = queryForSolrDocList(solrQuery, solrClient);
+            forEachDoc(docs, consumer);
+        } else {
+            System.out.println("NumFound is 0!");
```
tohle vubec neni treba, pokud nenajde zadne vysledky podle query, tak se vrati prazdny seznam a consumer nic neudela

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientApi.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public static void forEachDoc(SolrDocumentList docs, Consumer<SolrDocument> consumer) {
```
neni potreba, kdyz tak muzes vyuzit Stream API a aplikovat consumera
napr https://medium.com/swlh/understanding-java-8s-consumer-supplier-predicate-and-function-c1889b9423d

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientApi.java:
```diff
@@ -0,0 +1,64 @@
+        return solrClient.query(solrQuery).getResults();
+    }
+
+    public static void commitAndClose(SolrClient solrClient)
```
radeji prepis na `close(boolean commit)` a podle parametru commituj pred uzavrenim

Romx98 replied:
Tomuto trosku nerozumiem, cize ak som to dobre pochopil, tak ako parametre tam budu SolrClient a boolean commit. V tele metody budem kontrolovat, ci commit je true a na zaklade toho vyvolam close?

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientApi.java:
```diff
@@ -0,0 +1,64 @@
+import java.text.SimpleDateFormat;
+import java.util.function.Consumer;
+
+public class SolrClientApi {
```
proc se nazyva API? podle obsahu to je spis util trida

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.List;
+import java.util.function.Consumer;
+
+public class SolrVcRecords {
```
Trida ma strasne divny nazev, ktery rika nepravdu o tom, co uvnitr dela. Kdyz prectu `SolrVcRecords` tak bych osobne ocekaval, ze je to seznam Solr zaznamu, podobne jak to ma `SolrDocumentList`. Ale tady to uvnitr dela veskerou logiku, to by fakt chtelo aspon jiny nazev.

Romx98 replied:
S pomenovaním tried, metód,... je to dosť obtiažne :D

kazooo replied:
Máš pravdu, tohle se mi občas také nejde dobře :D ale tomu se naučiš s čásem

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        };
+    }
+
+    private SolrQuery createSolrQuery(String uuid) {
```
`createSolrQuery` nerika co to bude za query, prilis obecny nazev, co treba `createQueryByPidPath`?

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+    @SuppressWarnings("unchecked")
+    private Consumer<SolrDocument> createUnlinkConsumer(String idVc) {
+        return solrDoc -> {
+            List<String> collections = (List<String>) solrDoc.getFieldValue("collection");
```
`"collection"` stejne tak i `"PID"` do samostatneho souboru s konstantami

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        return solrDoc -> {
+            List<String> collections = (List<String>) solrDoc.getFieldValue("collection");
+
+            if ((collections != null) && (collections.contains(idVc))) {
```
zavorky nejsou potreba, prilis jednoduche podminky, ale jak chces

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        inputDoc.addField("PID", uuid);
+        inputDoc.addField("collection", Collections.singletonMap(modifier, collection));
+
+        inputDoc.addField("modified_date", Collections.singletonMap("set", dateFormat.format(new Date())));
```
`"modified_date"` take do konstant

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        modifyRecordInSolr(uuid, idVc, "set");
+    }
+
+    private void modifyRecordInSolr(String uuid, Object collection, String modifier)
```
`collection` muze byt napr `collections` nebo `collectionList`

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        modifyRecordInSolr(uuid, idVc, "set");
+    }
+
+    private void modifyRecordInSolr(String uuid, Object collection, String modifier)
```
`modifyRecordInSolr` je take obecny nazev. Kdyby tato funkce delala pouze inject objekt s danym modifikatorem do uuid, to by slo. Ale tvoje funkce navic setuje konkretne pole, to uz je neco jine

kazooo replied:
tady musis rozdelit inject objektu do SolrInputDocument podle klicu od zbyvajici logiky

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+        inputDoc.addField("collection", Collections.singletonMap(modifier, collection));
+
+        inputDoc.addField("modified_date", Collections.singletonMap("set", dateFormat.format(new Date())));
+        solrClient.add(inputDoc);
```
modifikujes zaznam a zaroven posilas do solru, takze funkce zaroven dela vic veci, tohle musis dekomponovat

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcRecords.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private SolrQuery createSolrQuery(String uuid) {
+        String allPartsQueryStr = "pid_path:/" + uuid.trim() + ".*/";
```
tady pouzij `rootPid`

kazooo replied:
protoze hledas podle korene vsech nasledovniku

kazooo reviewed (changes requested):
Commity a jejich popis vypadaji spatne, ve stromu commitu je videt jen `Summary description of changes:` a abys uvidel ce se tam provedlo musis kazdy popis rozklikavat. 

```
- designed for future use
 - !! ONLY removal VC !!
- tested on the local system, everything is running OK...
``` 
tohle se do popisu nepise

precti si nasledujici clanky:

https://www.freecodecamp.org/news/a-beginners-guide-to-git-how-to-write-a-good-commit-message/
https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/
https://chris.beams.io/posts/git-commit/

ten posledni je docela cool :D

Romx98 committed:
Modified and decompose code based on cr

Renaming class, methods and variables. Added new class `SolrField.java`,
contains string constants. Removal of the `forEachDoc` method and
replacement of foreach in `singleRequestAndApply` method.
Added newer version for Solrj in `build.gradle`.

Testing on the local system.

Romx98 committed:
MimeType fix for RELS-EXT datastream

Change mimeType in enum class `DataStreams` for RELS-EXT. Added
a new class that fix char `+` in all client request.
Added new method for configure rest template, set interceptors.

This change was due to the RestTemplate replacing `+` with a
space. Stackoverflow solution inspiration.

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/RemoveVcSolr.java:
```diff
@@ -0,0 +1,22 @@
+
+        SolrClient solrClient = SolrClientUtils.createSolrClient(solrHost);
+
+        SolrVcService solrVcRecords = new SolrVcService(solrClient);
```
solrVcRecords -> solrVcService

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientUtils.java:
```diff
@@ -0,0 +1,57 @@
+            throws SolrServerException, IOException {
+        long numFound = queryForNumFound(solrQuery, solrClient);
+
+        if(numFound != 0) {
```
proc tady potrebujes pocet dokumentu? i kdyz nebude zadny dokument, zadna zmena nenastane

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrClientUtils.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+public class SolrClientUtils {
```
tohle klidne muze byt jako utilita, ktera bere klienta a dela s nim nejake operace. Stejne tak muzes udelat wrapper pro klienta a pracovat s klientem zapouzdrene a izolovane.

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcService.java:
```diff
@@ -0,0 +1,73 @@
+        };
+    }
+
+    private SolrQuery createQueryByPidPath(String uuid) {
```
createQueryByPidPath -> createQueryByRootUuid

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcService.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private SolrQuery createQueryByPidPath(String uuid) {
+        String allPartsQueryStr = "root_pid:/" + uuid.trim() + ".*/";
```
u root_pid neni treba regex, dokonce to komplikuje hledani

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcService.java:
```diff
@@ -0,0 +1,73 @@
+    private void updateModifiedDateNow(SolrInputDocument inputDoc) {
+        inputDoc.addField(SolrField.MODIFIED_DATE, Collections.singletonMap("set", dateFormat.format(new Date())));
+    }
+}
```
dekomponuj trochu jinak
```java
public void setVcFor(final String uuid, final List<String> collections) {
    SolrInputDocument inputDoc = new SolrInputDocument();

    inputDoc.addField(SolrField.UUID, uuid);
    setSolrInputFieldValue(inputDoc, SolrField.COLLECTION, collections);
    // solr automaticky zkonvertuje, kdyztak to jeste zkontroluj
    setSolrInputFieldValue(inputDoc, SolrField.MODIFIED_DATE, new Date());

    solrClient.add(inputDoc);
}

public void setSolrInputFieldValue(final SolrInputDocument inputDoc, final String fieldName, final Object fieldValue) {
    inputDoc.addField(fieldName, Collections.singletonMap("set", fieldValue);
}
```

Romx98 replied:
Ano, nakoniec som si to overil a Solr si automaticky zkonvertuje format date.

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcService.java:
```diff
@@ -0,0 +1,73 @@
+                    String uuid = (String) solrDoc.getFieldValue(SolrField.UUID);
+                    setVcFor(uuid, collections);
+                } catch (SolrServerException | IOException e) {
+                    e.printStackTrace();
```
tady by chtelo nejakou informativnejsi hlasku, aspon u jakeho uuid nastal problem

kazooo reviewed (changes requested):
Commity vypadaji mnohem lip :+1: Jenom prosim delej commit pro kazdou zmenu v ramci tasku, ktery resis. Napr. jeden commit pro zmenu mimeType u  DataStreams, dalsi commit pridani interceptora atd... Je lepsi kdyz je vic malych commitu, nez nekolik obrovskych. Kazdopadne na konci udelas squash a veskere funkcionalni zmeny budou na jednom miste.

Romx98 committed:
Modified code based on cr

Renaming a variable 'solrVcRecords' -> 'solrVcService' and
method in SolrVcService.
Remove query for finding number of documents.

Romx98 committed:
Decompose method and modify class

Added statement when invoking an exception in 'createUnlinkConsumer'
Decompose method setVcFor, method renaming for set field value
in solr. Removed method and body added to method setVcFor.
Editing string for solr query ('root_pid').

Tested.

kazooo commented on src/main/java/cz/mzk/fedora/solrcorrection/SolrVcService.java:
```diff
@@ -35,39 +34,37 @@ public void removeVc(String idVc, String uuid) throws SolrServerException, IOExc
         inputDoc.addField(SolrField.UUID, uuid);
-        inputDoc.addField(SolrField.COLLECTION, Collections.singletonMap(modifier, collectionList));
+        setSolrInInFieldValue(inputDoc, SolrField.COLLECTION, collectionList);
+        setSolrInInFieldValue(inputDoc, SolrField.MODIFIED_DATE, new Date());
```
neni uplne dobry nazev, tam bych dal treba `updateSolrFieldValue`, co myslis?

kazooo reviewed (changes requested):
Dobre, kdyz to funguje tak super! Jen zmen nazev metody a udelame tady squash aby vsechno vypadalo ciste a krasne :)

Romx98 committed:
Method renaming
